libs/embassy-container-init/src/main.rs

```diff
@@ -0,0 +1,239 @@
+use std::{collections::BTreeMap, process::Stdio, sync::Arc};
+
+use async_stream::stream;
+use color_eyre::Report;
+use futures::{pin_mut, Stream, StreamExt};
+use tokio::{
+    io::AsyncBufReadExt,
+    select,
+    sync::{oneshot, Mutex},
+};
+use tokio::{io::BufReader, process::Command};
+use tracing::instrument;
+
+use embassy_container_init::{Input, InputJsonRpc, JsonRpc, Output, OutputJsonRpc, RpcId};
+
+const MAX_COMMANDS: usize = 10;
+
+#[derive(Debug, Clone)]
+struct Io {
+    commands: Arc<Mutex<BTreeMap<RpcId, oneshot::Sender<()>>>>,
+    ids: Arc<Mutex<BTreeMap<RpcId, u32>>>,
+}
+
+impl Io {
+    async fn clean_id(&self, id: &RpcId) -> (Option<u32>, Option<oneshot::Sender<()>>) {
+        (
+            self.ids.lock().await.remove(id),
+            self.commands.lock().await.remove(id),
+        )
+    }
+    async fn trigger_end_command(&self, id: RpcId) {
+        if let Some(command) = self.commands.lock().await.remove(&id) {
+            if command.send(()).is_err() {
+                tracing::trace!("Command {id:?} could not be ended, possible error or was done");
+            }
+        }
+    }
+
+    async fn create_end_command(&self, id: RpcId) -> oneshot::Receiver<()> {
+        let (send, receiver) = oneshot::channel();
+        if let Some(other_command) = self.commands.lock().await.insert(id.clone(), send) {
+            if other_command.send(()).is_err() {
+                tracing::trace!(
+                    "Found other command {id:?} could not be ended, possible error or was done"
+                );
+            }
+        }
+        receiver
+    }
+
+    fn start() -> Self {
+        use tracing_error::ErrorLayer;
+        use tracing_subscriber::prelude::*;
+        use tracing_subscriber::{fmt, EnvFilter};
+
+        let filter_layer = EnvFilter::new("embassy_container_init=trace");
+        let fmt_layer = fmt::layer().with_target(true);
+
+        tracing_subscriber::registry()
+            .with(filter_layer)
+            .with(fmt_layer)
+            .with(ErrorLayer::default())
+            .init();
+        color_eyre::install().unwrap();
+        Self {
+            commands: Default::default(),
+            ids: Default::default(),
+        }
+    }
+
+    #[instrument]
+    fn command(&self, input: InputJsonRpc) -> impl Stream<Item = OutputJsonRpc> {
+        let io = self.clone();
+        stream! {
+            let (id, command) = input.into_pair();
+            match command {
```

Comment:
@dr-bonez  :question: Should this be extracted to another function to create consistent abstraction levels?